GabijaBern/assignment2/BinaryTreeAncestors.java

```diff
@@ -0,0 +1,251 @@
+/**
+*Class BinaryTreeAncestors is the solution for the first question in second assignment 
+*Here I have implemented my own tree where only unique values are available
+*/
+import java.util.Queue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+
+public class BinaryTreeAncestors
+{
+	public static void main(String[] args)
+	{
+		BTree<Integer> test = new BTree<Integer>();
+		test.treeRoot = new Node<Integer>(0);
+		test.insert(test.treeRoot, 1);
+		test.insert(test.treeRoot, 2);
+		test.insert(test.treeRoot, 3);
+		test.insert(test.treeRoot, 4);
+		test.insert(test.treeRoot, 5);
+		test.insert(test.treeRoot, 6);
+		
+		//test.findAncestors(test, new Node<Integer>(3));
+		
+		//Node<Integer> node = new Node<Integer>(null);
+		//node = test.findCommonAncestor(test, test.treeRoot.left.treeRoot.left.treeRoot, test.treeRoot.right.treeRoot);
+		//System.out.println(node);
+		//System.out.println(test.findCommonAncestor2(test, test.treeRoot.left.treeRoot.left.treeRoot, test.treeRoot.right.treeRoot.right.treeRoot));
+	}
+}
+/*
+ * Generic class for Nodes
+ * 
+ * Here we override toString that is used while testing the methods
+ */
+class Node<T> 
+{
+	public T value;
+	public BTree<T> left = null, right = null;
+	
+	public Node(T valueToSet)
+	{
+		this.value = valueToSet;
+	}
+	
+	@Override
+	public String toString()
+	{
+		return "" + value;
+	}
+}
+/*
+ * Binary tree implementation with insertion and ancestors methods
+ */
+class BTree<T>
+{
+	Node<T> treeRoot = null;
+	
+	/*
+	 * Method for filling levels of tree
+	 * we keep a queue of nodes that don't have two children yet, every time we add a node
+	 * we add it to the end of the queue and make it a child of a first node in a queue
+	 * once the node in front has to children, we remove the node from the list
+	 * 
+	 * @param tree The tree into which we want to insert a new node
+	 * @param valueToInsert The value that we want to insert
+	 * 
+	 * @throws NoSuchElementException Exception can be thrown if queue is empty
+	 */
+	public void insert(Node<T> tree, T valueToInsert) throws NoSuchElementException
+	{
+		Queue<Node<T>> nodesQueue = new LinkedList<Node<T>>();
+        nodesQueue.add(tree);
+ 
+        while (!nodesQueue.isEmpty()) 
+        {
+        	//we take first node that doesn't have two children in the list
+        	Node<T> firstNode = nodesQueue.peek();
+        	nodesQueue.remove();
+        	
+        	//we start filling from left to right
+            if (firstNode.left == null)
+            {
+            	firstNode.left = new BTree<T>();
+                firstNode.left.treeRoot = new Node<T>(valueToInsert);
+                break;
+            }
+            else
+            	nodesQueue.add(firstNode.left.treeRoot);

```

Comment:
The queue only ever has length 0 or 1 doesn't it? So you don't need a queue, you could do all this with a variable.

Reply:
No, the size can be bigger than it and we do need the queue. My comments might have been unclear, I have updated them now, sorry. :)